File: linux/cpp/src/latency.cpp
#include "fredplayer/latency.hpp"

#include <gst/app/gstappsink.h>
#include <gst/app/gstappsrc.h>
#include <gst/audio/gstaudiobasesink.h>
#include <gst/gst.h>
#include <fftw3.h>
#include <nlohmann/json.hpp>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <memory>
#include <stdexcept>
#include <thread>

namespace fredplayer {
namespace {
using json = nlohmann::json;
constexpr int kRate = 48000;
constexpr double kPi = 3.14159265358979323846;

json commandJson(const char* command) {
  FILE* pipe = popen(command, "r");
  if (!pipe) return {};
  std::string output; std::array<char, 4096> buffer{};
  while (fgets(buffer.data(), buffer.size(), pipe)) output += buffer.data();
  pclose(pipe);
  try { return json::parse(output); } catch (...) { return {}; }
}

GstElement* make(const char* factory, const char* name) {
  auto* element = gst_element_factory_make(factory, name);
  if (!element) throw std::runtime_error(std::string("Missing GStreamer element: ") + factory);
  return element;
}

GstElement* outputSink(const AudioOutput& output, const char* name) {
  auto* sink = gst_element_factory_make("pulsesink", name);
  if (sink && !output.key.empty()) g_object_set(sink, "device", output.key.c_str(), nullptr);
  if (!sink) sink = make("autoaudiosink", name);
  return sink;
}

std::vector<float> chirp() {
  const int count = static_cast<int>(.30 * kRate);
  std::vector<float> result(count);
  double phase = 0;
  for (int i = 0; i < count; ++i) {
    const auto t = static_cast<double>(i) / count;
    const auto frequency = 700.0 * std::pow(5600.0 / 700.0, t);
    phase += 2.0 * kPi * frequency / kRate;
    const auto envelope = .5 - .5 * std::cos(2.0 * kPi * i / (count - 1));
    result[i] = static_cast<float>(.32 * envelope * std::sin(phase));
  }
  return result;
}

std::pair<int, double> correlate(const std::vector<float>& capture,
                                 const std::vector<float>& pattern) {
  if (capture.size() < pattern.size()) return {-1, 0};
  std::size_t fftSize = 1;
  while (fftSize < capture.size() + pattern.size()) fftSize <<= 1;
  std::vector<float> left(fftSize), right(fftSize), inverse(fftSize);
  std::copy(capture.begin(), capture.end(), left.begin());
  std::reverse_copy(pattern.begin(), pattern.end(), right.begin());
  std::vector<fftwf_complex> lf(fftSize / 2 + 1), rf(fftSize / 2 + 1), product(fftSize / 2 + 1);
  auto lp = fftwf_plan_dft_r2c_1d(fftSize, left.data(), lf.data(), FFTW_ESTIMATE);
  auto rp = fftwf_plan_dft_r2c_1d(fftSize, right.data(), rf.data(), FFTW_ESTIMATE);
  fftwf_execute(lp); fftwf_execute(rp);
  for (std::size_t i = 0; i < product.size(); ++i) {
    product[i][0] = lf[i][0] * rf[i][0] - lf[i][1] * rf[i][1];
    product[i][1] = lf[i][0] * rf[i][1] + lf[i][1] * rf[i][0];
  }
  auto ip = fftwf_plan_dft_c2r_1d(fftSize, product.data(), inverse.data(), FFTW_ESTIMATE);
  fftwf_execute(ip); fftwf_destroy_plan(lp); fftwf_destroy_plan(rp); fftwf_destroy_plan(ip);
  const auto first = pattern.size() - 1;
  const auto last = first + capture.size() - pattern.size();
  auto found = std::max_element(inverse.begin() + first, inverse.begin() + last + 1,
                                [](float a, float b) { return std::abs(a) < std::abs(b); });
  const auto match = static_cast<int>(std::distance(inverse.begin(), found) - first);
  double patternPower = 0, capturePower = 0;
  for (std::size_t i = 0; i < pattern.size(); ++i) {
    patternPower += pattern[i] * pattern[i];
    capturePower += capture[match + i] * capture[match + i];
  }
  const auto raw = std::abs(*found) / fftSize;
  const auto confidence = raw / std::sqrt(std::max(1e-12, patternPower * capturePower));
  return {match, confidence};
}

}  // namespace

AudioOutput currentOutput() {
  const auto info = commandJson("pactl -f json info 2>/dev/null");
  const auto key = info.value("default_sink_name", "");
  const auto sinks = commandJson("pactl -f json list sinks 2>/dev/null");
  if (sinks.is_array()) for (const auto& sink : sinks) {
    if (sink.value("name", "") != key) continue;
    const auto properties = sink.value("properties", json::object());
    const auto bluetooth = key.rfind("bluez_output.", 0) == 0 || properties.value("device.bus", "") == "bluetooth";
    return {key, sink.value("description", key), bluetooth};
  }
  return {key, key.empty() ? "Default audio output" : key, key.rfind("bluez_output.", 0) == 0};
}

std::vector<Microphone> microphones() {
  const auto info = commandJson("pactl -f json info 2>/dev/null");
  const auto defaultKey = info.value("default_source_name", "");
  std::vector<Microphone> result;
  const auto sources = commandJson("pactl -f json list sources 2>/dev/null");
  if (sources.is_array()) for (const auto& source : sources) {
    const auto key = source.value("name", "");
    if (key.empty() || (key.size() >= 8 && key.compare(key.size() - 8, 8, ".monitor") == 0)) continue;
    result.push_back({key, source.value("description", key), key == defaultKey});
  }
  std::sort(result.begin(), result.end(), [](const auto& a, const auto& b) {
    return a.isDefault != b.isDefault ? a.isDefault : a.label < b.label;
  });
  return result;
}

int probeSystemLatency(const AudioOutput& output) {
  GstElement* pipeline = gst_pipeline_new("fred-native-latency-probe");
  auto* source = make("audiotestsrc", "latency-silence");
  auto* convert = make("audioconvert", "latency-convert");
  auto* sink = outputSink(output, "latency-output");
  g_object_set(source, "is-live", TRUE, "wave", 4, nullptr);
  gst_bin_add_many(GST_BIN(pipeline), source, convert, sink, nullptr);
  if (!gst_element_link_many(source, convert, sink, nullptr)) throw std::runtime_error("Could not build latency probe");
  gst_element_set_state(pipeline, GST_STATE_PLAYING);
  gst_element_get_state(pipeline, nullptr, nullptr, 3 * GST_SECOND);
  GstQuery* query = gst_query_new_latency();
  GstClockTime minimum = 0, maximum = 0; gboolean live = false;
  if (!gst_element_query(pipeline, query)) minimum = 0;
  else gst_query_parse_latency(query, &live, &minimum, &maximum);
  gst_query_unref(query);
  if (GST_IS_BASE_SINK(sink))
    minimum = std::max(minimum, gst_base_sink_get_latency(GST_BASE_SINK(sink)));
  gst_element_set_state(pipeline, GST_STATE_NULL); gst_object_unref(pipeline);
  if (!minimum) throw std::runtime_error("The audio system did not report output latency");
  return std::clamp(static_cast<int>(std::llround(minimum / 1'000'000.0)), 0, 1500);
}

CalibrationResult calibrateWithMicrophone(const std::string& microphoneKey) {
  if (microphoneKey.empty()) throw std::runtime_error("Choose a microphone first");
  const auto activeOutput = currentOutput();
  GstElement* pipeline = gst_pipeline_new("fred-native-speaker-calibration");
  auto* source = make("pulsesrc", "calibration-microphone");
  g_object_set(source, "device", microphoneKey.c_str(), nullptr);
  auto* inConvert = make("audioconvert", "calibration-in-convert");
  auto* inResample = make("audioresample", "calibration-in-resample");
  auto* inCaps = make("capsfilter", "calibration-in-caps");
  auto* sink = make("appsink", "calibration-capture");
  auto* appSource = make("appsrc", "calibration-sound");
  auto* outConvert = make("audioconvert", "calibration-out-convert");
  auto* outResample = make("audioresample", "calibration-out-resample");
  auto* audioSink = outputSink(activeOutput, "calibration-output");
  auto* caps = gst_caps_from_string("audio/x-raw,format=F32LE,channels=1,rate=48000,layout=interleaved");
  g_object_set(inCaps, "caps", caps, nullptr);
  g_object_set(appSource, "caps", caps, "format", GST_FORMAT_TIME, "is-live", TRUE, "block", TRUE, nullptr);
  gst_caps_unref(caps);
  g_object_set(sink, "sync", FALSE, "max-buffers", 256, "drop", FALSE, nullptr);
  gst_bin_add_many(GST_BIN(pipeline), source, inConvert, inResample, inCaps, sink,
                   appSource, outConvert, outResample, audioSink, nullptr);
  if (!gst_element_link_many(source, inConvert, inResample, inCaps, sink, nullptr) ||
      !gst_element_link_many(appSource, outConvert, outResample, audioSink, nullptr))
    throw std::runtime_error("Could not build speaker calibration pipeline");

  auto pattern = chirp();
  const int preRoll = static_cast<int>(.4 * kRate);
  std::vector<float> outputSamples(preRoll + pattern.size() + static_cast<int>(.35 * kRate));
  std::copy(pattern.begin(), pattern.end(), outputSamples.begin() + preRoll);
  std::vector<float> captured;
  gst_element_set_state(pipeline, GST_STATE_PLAYING);
  gst_element_get_state(pipeline, nullptr, nullptr, 3 * GST_SECOND);
  GstBuffer* buffer = gst_buffer_new_allocate(nullptr, outputSamples.size() * sizeof(float), nullptr);
  gst_buffer_fill(buffer, 0, outputSamples.data(), outputSamples.size() * sizeof(float));
  GST_BUFFER_DURATION(buffer) = outputSamples.size() * GST_SECOND / kRate;
  if (gst_app_src_push_buffer(GST_APP_SRC(appSource), buffer) != GST_FLOW_OK)
    throw std::runtime_error("Could not play calibration sound");
  const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(3);
  while (std::chrono::steady_clock::now() < deadline) {
    GstSample* sample = gst_app_sink_try_pull_sample(GST_APP_SINK(sink), 100 * GST_MSECOND);
    if (!sample) continue;
    GstBuffer* received = gst_sample_get_buffer(sample); GstMapInfo map{};
    if (gst_buffer_map(received, &map, GST_MAP_READ)) {
      const auto* values = reinterpret_cast<const float*>(map.data);
      captured.insert(captured.end(), values, values + map.size / sizeof(float));
      gst_buffer_unmap(received, &map);
    }
    gst_sample_unref(sample);
  }
  gst_app_src_end_of_stream(GST_APP_SRC(appSource));
  gst_element_set_state(pipeline, GST_STATE_NULL); gst_object_unref(pipeline);
  if (currentOutput().key != activeOutput.key) throw std::runtime_error("The audio output changed during calibration");
  const auto [match, confidence] = correlate(captured, pattern);
  if (match < 0 || confidence < .10) throw std::runtime_error("The microphone could not hear the calibration sound clearly");
  const auto delay = std::clamp(static_cast<int>(std::llround((match - preRoll) * 1000.0 / kRate)), 0, 1500);
  return {activeOutput, delay, confidence};
}

}  // namespace fredplayer

File: linux/fredplayer/latency.py
from __future__ import annotations

from dataclasses import dataclass
import json
import subprocess
import time
from typing import Optional

import gi

try:
    import numpy as np
except Exception:  # pragma: no cover - optional at runtime
    np = None

gi.require_version("Gst", "1.0")
gi.require_version("GstAudio", "1.0")
gi.require_version("GstApp", "1.0")
from gi.repository import Gst, GstAudio  # noqa: E402


Gst.init(None)

SAMPLE_RATE = 48_000
PRE_ROLL_SECONDS = 0.40
CHIRP_SECONDS = 0.30
POST_ROLL_SECONDS = 0.35
CAPTURE_AFTER_CHIRP_SECONDS = 1.75
MAX_DELAY_MS = 1_500
MIN_CONFIDENCE = 0.10
PULL_TIMEOUT_NS = 100 * Gst.MSECOND
MONO_CAPS = Gst.Caps.from_string(
    f"audio/x-raw,format=F32LE,channels=1,rate={SAMPLE_RATE},layout=interleaved"
)


@dataclass(frozen=True)
class AudioOutput:
    key: str
    label: str
    bluetooth: bool


@dataclass(frozen=True)
class Microphone:
    key: str
    label: str
    default: bool = False


@dataclass(frozen=True)
class CalibrationResult:
    output: AudioOutput
    delay_ms: int
    confidence: float


def current_output() -> AudioOutput:
    default_name = _pactl_default("sink")
    sinks = _pactl_items("sinks")
    sink = next((item for item in sinks if str(item.get("name", "")) == default_name), None)
    if sink is None:
        return AudioOutput(default_name, default_name or "Default audio output", False)

    name = str(sink.get("name", "")) or default_name
    label = str(sink.get("description", "")).strip() or name or "Default audio output"
    properties = sink.get("properties", {})
    bluetooth = name.startswith("bluez_output.") or (
        isinstance(properties, dict)
        and str(properties.get("device.bus", "")).casefold() == "bluetooth"
    )
    return AudioOutput(name, label, bluetooth)


def microphones() -> list[Microphone]:
    default_name = _pactl_default("source")
    result: list[Microphone] = []
    for source in _pactl_items("sources"):
        name = str(source.get("name", ""))
        if not name or name.endswith(".monitor"):
            continue
        label = str(source.get("description", "")).strip() or name
        result.append(Microphone(name, label, name == default_name))
    result.sort(key=lambda item: (not item.default, item.label.casefold(), item.key))
    return result


def probe_system_latency(output: AudioOutput) -> int:
    """Ask the active GStreamer audio sink for its system-reported latency.

    A live silent source is intentional. GstAudioBaseSink only exposes a useful
    value after the selected output has negotiated a live stream; a non-live
    file-player source commonly reports zero even for Bluetooth.
    """
    pipeline = Gst.Pipeline.new("fred-system-latency-probe")
    source = _make_element("audiotestsrc", "latency-silence")
    convert = _make_element("audioconvert", "latency-convert")
    resample = _make_element("audioresample", "latency-resample")
    sink = _make_output_sink(output, "latency-output")
    source.set_property("is-live", True)
    source.set_property("wave", 4)  # GstAudioTestSrcWave.SILENCE

    for element in (source, convert, resample, sink):
        pipeline.add(element)
    if not source.link(convert) or not convert.link(resample) or not resample.link(sink):
        raise RuntimeError("Could not build the system latency probe")

    try:
        pipeline.set_state(Gst.State.PLAYING)
        pipeline.get_state(3 * Gst.SECOND)
        deadline = time.monotonic() + 1.5
        latency_ns = 0
        while time.monotonic() < deadline and latency_ns <= 0:
            _raise_bus_error(pipeline.get_bus())
            latency_ns = _audio_sink_latency_ns(sink)
            if latency_ns <= 0:
                time.sleep(0.08)
        if latency_ns <= 0:
            query = Gst.Query.new_latency()
            if pipeline.query(query):
                _live, minimum, _maximum = query.parse_latency()
                latency_ns = max(0, int(minimum))
        if latency_ns <= 0:
            raise RuntimeError("The audio system did not report output latency")
        return max(0, min(MAX_DELAY_MS, round(latency_ns / Gst.MSECOND)))
    finally:
        pipeline.set_state(Gst.State.NULL)


def calibrate_with_microphone(microphone_key: str) -> CalibrationResult:
    if np is None:
        raise RuntimeError("Microphone calibration requires python3-numpy")
    if not microphone_key:
        raise RuntimeError("Choose a microphone first")

    output = current_output()
    if not output.key:
        raise RuntimeError("No audio output is available")

    pipeline = Gst.Pipeline.new("fred-speaker-latency-calibration")
    source = _make_element("pulsesrc", "calibration-microphone")
    source.set_property("device", microphone_key)
    input_convert = _make_element("audioconvert", "calibration-input-convert")
    input_resample = _make_element("audioresample", "calibration-input-resample")
    input_caps = _make_element("capsfilter", "calibration-input-caps")
    input_caps.set_property("caps", MONO_CAPS)
    appsink = _make_element("appsink", "calibration-capture")
    appsink.set_property("emit-signals", False)
    appsink.set_property("sync", False)
    appsink.set_property("max-buffers", 256)
    appsink.set_property("drop", False)

    appsrc = _make_element("appsrc", "calibration-sound")
    appsrc.set_property("caps", MONO_CAPS)
    appsrc.set_property("format", Gst.Format.TIME)
    appsrc.set_property("is-live", True)
    appsrc.set_property("block", True)
    output_queue = _make_element("queue", "calibration-output-queue")
    output_convert = _make_element("audioconvert", "calibration-output-convert")
    output_resample = _make_element("audioresample", "calibration-output-resample")
    output_sink = _make_output_sink(output, "calibration-output")

    elements = (
        source,
        input_convert,
        input_resample,
        input_caps,
        appsink,
        appsrc,
        output_queue,
        output_convert,
        output_resample,
        output_sink,
    )
    for element in elements:
        pipeline.add(element)
    if (
        not source.link(input_convert)
        or not input_convert.link(input_resample)
        or not input_resample.link(input_caps)
        or not input_caps.link(appsink)
        or not appsrc.link(output_queue)
        or not output_queue.link(output_convert)
        or not output_convert.link(output_resample)
        or not output_resample.link(output_sink)
    ):
        raise RuntimeError("Could not build the speaker calibration pipeline")

    captured: list[tuple[int, "np.ndarray"]] = []
    chirp = _build_chirp()
    output_samples = np.concatenate(
        (
            np.zeros(round(PRE_ROLL_SECONDS * SAMPLE_RATE), dtype="<f4"),
            chirp,
            np.zeros(round(POST_ROLL_SECONDS * SAMPLE_RATE), dtype="<f4"),
        )
    )

    try:
        pipeline.set_state(Gst.State.PLAYING)
        first_sample = _pull_sample(appsink, pipeline, 3.0)
        if first_sample is None:
            raise RuntimeError("The selected microphone did not provide audio")
        _append_capture(captured, first_sample)

        clock = pipeline.get_clock()
        if clock is None:
            raise RuntimeError("The audio system did not provide a shared clock")
        running_time = max(0, int(clock.get_time() - pipeline.get_base_time()))
        scheduled_start_ns = running_time + 120 * Gst.MSECOND
        expected_chirp_ns = scheduled_start_ns + round(PRE_ROLL_SECONDS * Gst.SECOND)

        buffer = Gst.Buffer.new_allocate(None, output_samples.nbytes, None)
        buffer.fill(0, output_samples.tobytes())
        buffer.pts = scheduled_start_ns
        buffer.dts = scheduled_start_ns
        buffer.duration = round(len(output_samples) * Gst.SECOND / SAMPLE_RATE)
        result = appsrc.emit("push-buffer", buffer)
        if result != Gst.FlowReturn.OK:
            raise RuntimeError("Could not play the calibration sound")

        stop_running_time = expected_chirp_ns + round(CAPTURE_AFTER_CHIRP_SECONDS * Gst.SECOND)
        wall_deadline = time.monotonic() + 5.0
        while time.monotonic() < wall_deadline:
            _raise_bus_error(pipeline.get_bus())
            sample = appsink.emit("try-pull-sample", PULL_TIMEOUT_NS)
            if sample is not None:
                _append_capture(captured, sample)
            now_running = max(0, int(clock.get_time() - pipeline.get_base_time()))
            if now_running >= stop_running_time:
                break
        appsrc.emit("end-of-stream")

        if current_output().key != output.key:
            raise RuntimeError("The audio output changed during calibration")
        capture, capture_start_ns = _capture_timeline(captured)
        match_frame, confidence = _find_chirp(capture, chirp, expected_chirp_ns, capture_start_ns)
        delay_ms = round(
            ((capture_start_ns + match_frame * Gst.SECOND / SAMPLE_RATE) - expected_chirp_ns)
            / Gst.MSECOND
        )
        if confidence < MIN_CONFIDENCE:
            raise RuntimeError(
                "The calibration sound was not clear enough; move the microphone closer or raise the speaker volume"
            )
        if delay_ms < 0 or delay_ms > MAX_DELAY_MS:
            raise RuntimeError("The measured speaker delay was outside the supported range")
        return CalibrationResult(output, int(delay_ms), float(confidence))
    finally:
        pipeline.set_state(Gst.State.NULL)


def _build_chirp() -> "np.ndarray":
    frames = round(CHIRP_SECONDS * SAMPLE_RATE)
    times = np.arange(frames, dtype=np.float64) / SAMPLE_RATE
    start_hz = 700.0
    end_hz = 6_500.0
    sweep = (end_hz - start_hz) / CHIRP_SECONDS
    phase = 2.0 * np.pi * (start_hz * times + 0.5 * sweep * times * times)
    envelope = np.hanning(frames)
    return (np.sin(phase) * envelope * 0.22).astype("<f4")


def _find_chirp(
    capture: "np.ndarray",
    chirp: "np.ndarray",
    expected_chirp_ns: int,
    capture_start_ns: int,
) -> tuple[int, float]:
    downsample = 8
    capture_ds = np.asarray(capture[::downsample], dtype=np.float64)
    reference = np.asarray(chirp[::downsample], dtype=np.float64)
    expected_frame = round((expected_chirp_ns - capture_start_ns) * SAMPLE_RATE / Gst.SECOND)
    earliest_frame = max(0, expected_frame - round(0.20 * SAMPLE_RATE))
    latest_frame = min(
        max(0, len(capture) - len(chirp)),
        expected_frame + round(1.50 * SAMPLE_RATE),
    )
    first = earliest_frame // downsample
    last = latest_frame // downsample
    search = capture_ds[first : last + len(reference)]
    if len(search) < len(reference):
        raise RuntimeError("Not enough microphone audio was captured")

    correlation = np.correlate(search, reference, mode="valid")
    reference_energy = float(np.dot(reference, reference))
    squared = search * search
    cumulative = np.concatenate(([0.0], np.cumsum(squared)))
    window_energy = cumulative[len(reference) :] - cumulative[: -len(reference)]
    denominator = np.sqrt(np.maximum(1e-20, reference_energy * window_energy))
    scores = np.abs(correlation) / denominator
    best = int(np.argmax(scores))
    return (first + best) * downsample, float(scores[best])


def _capture_timeline(captured: list[tuple[int, "np.ndarray"]]) -> tuple["np.ndarray", int]:
    valid = [(pts, samples) for pts, samples in captured if pts >= 0 and len(samples)]
    if not valid:
        raise RuntimeError("No usable microphone audio was captured")
    start_ns = min(pts for pts, _samples in valid)
    end_frame = max(
        round((pts - start_ns) * SAMPLE_RATE / Gst.SECOND) + len(samples)
        for pts, samples in valid
    )
    timeline = np.zeros(max(1, end_frame), dtype="<f4")
    for pts, samples in valid:
        start = max(0, round((pts - start_ns) * SAMPLE_RATE / Gst.SECOND))
        end = min(len(timeline), start + len(samples))
        if end > start:
            timeline[start:end] = samples[: end - start]
    return timeline, start_ns


def _append_capture(captured: list[tuple[int, "np.ndarray"]], sample: Gst.Sample) -> None:
    buffer = sample.get_buffer()
    if buffer is None:
        return
    ok, info = buffer.map(Gst.MapFlags.READ)
    if not ok:
        return
    try:
        samples = np.frombuffer(info.data, dtype="<f4").copy()
    finally:
        buffer.unmap(info)
    pts = int(buffer.pts) if buffer.pts != Gst.CLOCK_TIME_NONE else -1
    if pts < 0 and captured:
        previous_pts, previous = captured[-1]
        pts = previous_pts + round(len(previous) * Gst.SECOND / SAMPLE_RATE)
    elif pts < 0:
        pts = 0
    captured.append((pts, samples))


def _pull_sample(appsink: Gst.Element, pipeline: Gst.Pipeline, timeout_seconds: float) -> Optional[Gst.Sample]:
    deadline = time.monotonic() + timeout_seconds
    while time.monotonic() < deadline:
        _raise_bus_error(pipeline.get_bus())
        sample = appsink.emit("try-pull-sample", PULL_TIMEOUT_NS)
        if sample is not None:
            return sample
    return None


def _make_output_sink(output: AudioOutput, name: str) -> Gst.Element:
    sink = Gst.ElementFactory.make("pulsesink", name)
    if sink is not None:
        if output.key:
            sink.set_property("device", output.key)
        return sink
    sink = Gst.ElementFactory.make("autoaudiosink", name)
    if sink is None:
        raise RuntimeError("No GStreamer audio output is available")
    return sink


def _audio_sink_latency_ns(sink: Gst.Element) -> int:
    candidates = [sink]
    if hasattr(sink, "get_children_count"):
        try:
            candidates.extend(sink.get_child_by_index(index) for index in range(sink.get_children_count()))
        except Exception:
            pass
    for candidate in candidates:
        try:
            latency = int(GstAudio.AudioBaseSink.get_latency(candidate))
        except Exception:
            continue
        if latency > 0:
            return latency
    return 0


def _make_element(factory: str, name: str) -> Gst.Element:
    element = Gst.ElementFactory.make(factory, name)
    if element is None:
        raise RuntimeError(f"Missing GStreamer element: {factory}")
    return element


def _raise_bus_error(bus: Gst.Bus) -> None:
    while True:
        message = bus.pop_filtered(Gst.MessageType.ERROR)
        if message is None:
            return
        error, _debug = message.parse_error()
        raise RuntimeError(error.message)


def _pactl_default(kind: str) -> str:
    command = "get-default-sink" if kind == "sink" else "get-default-source"
    try:
        result = subprocess.run(
            ["pactl", command],
            check=True,
            capture_output=True,
            text=True,
            timeout=2.0,
        )
        return result.stdout.strip()
    except (OSError, subprocess.SubprocessError):
        return ""


def _pactl_items(kind: str) -> list[dict]:
    try:
        result = subprocess.run(
            ["pactl", "-f", "json", "list", kind],
            check=True,
            capture_output=True,
            text=True,
            timeout=2.0,
        )
        value = json.loads(result.stdout)
        return [item for item in value if isinstance(item, dict)] if isinstance(value, list) else []
    except (OSError, subprocess.SubprocessError, json.JSONDecodeError):
        return []
